File: app/src/main/java/ru/igla/tfprofiler/utils/IOUtils.java
package ru.igla.tfprofiler.utils;

import android.content.Context;
import android.os.Environment;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.Closeable;
import java.io.File;


public final class IOUtils {

    private IOUtils() {
        //ignore
    }

    private static String combinePaths(@NonNull String... paths) {
        if (paths.length == 0) return "";
        File file = new File(paths[0]);
        for (int i = 1; i < paths.length; i++) {
            file = new File(file, paths[i]);
        }
        return file.getPath();
    }

    private static String getFilesDirAbsolutePath(Context context, String basePath) {
        return combinePaths(
                basePath,
                "/Android/data/",
                context.getPackageName(),
                "/files/"
        );
    }

    @Nullable
    private static String getExternalFileDir(Context c) {
        File file = c.getExternalFilesDir(null);
        if (file == null) {
            return null;
        }
        String path = file.getAbsolutePath(); //we can receive with missing '/' in the end of path
        if (!path.endsWith("/")) {
            return path + "/";
        }
        return path; //return file sdcard directory
    }

    public static String getAppSpecificFilesDirAbsolutePath(Context context) {
        String externalDir = getExternalFileDir(context);
        if (StringUtils.isNullOrEmpty(externalDir)) {
            return getFilesDirAbsolutePath(context, getBaseExternalStoragePath());
        }
        return externalDir;
    }

    private static String getBaseExternalStoragePath() {
        return Environment.getExternalStorageDirectory().getAbsolutePath();
    }

    /**
     * Closes 'closeable', ignoring any checked exceptions. Does nothing if 'closeable' is null.
     */
    public static void closeQuietly(@Nullable final Closeable closeable) {
        if (closeable != null) {
            try {
                closeable.close();
            } catch (Exception ignored) {
                //ignore
            }
        }
    }
}

File: app/src/main/java/ru/igla/tfprofiler/utils/IOUtils.kt
package ru.igla.tfprofiler.utils

import android.content.Context
import android.os.Environment
import ru.igla.tfprofiler.utils.StringUtils.isNullOrEmpty
import java.io.Closeable
import java.io.File

object IOUtils {

    private fun combinePaths(vararg paths: String): String {
        if (paths.isEmpty()) return ""
        var file = File(paths[0])
        for (i in 1 until paths.size) {
            file = File(file, paths[i])
        }
        return file.path
    }

    private fun getFilesDirAbsolutePath(context: Context, basePath: String): String {
        return combinePaths(
            basePath,
            "/Android/data/",
            context.packageName,
            "/files/"
        )
    }

    private fun getExternalFileDir(c: Context): String? {
        val file = c.getExternalFilesDir(null) ?: return null
        val path = file.absolutePath //we can receive with missing '/' in the end of path
        return if (!path.endsWith("/")) {
            "$path/"
        } else path
        //return file sdcard directory
    }

    fun getAppSpecificFilesDirAbsolutePath(context: Context): String? {
        val externalDir = getExternalFileDir(context)
        return if (isNullOrEmpty(externalDir)) {
            getFilesDirAbsolutePath(
                context,
                baseExternalStoragePath
            )
        } else externalDir
    }

    private val baseExternalStoragePath: String
        get() = Environment.getExternalStorageDirectory().absolutePath

    /**
     * Closes 'closeable', ignoring any checked exceptions. Does nothing if 'closeable' is null.
     */
    @JvmStatic
    fun closeQuietly(closeable: Closeable?) {
        closeable?.let {
            try {
                it.close()
            } catch (ignored: Exception) {
                //ignore
            }
        }
    }
}